src/w3d/renderer/assetmgr.cpp

```diff
@@ -816,6 +896,89 @@ bool GameAssetManager::Recolour_Asset(RenderObjClass *robj, uint32_t colour)
     }
 }
 
+TextureClass *GameAssetManager::Recolor_Texture(TextureClass *texture, int color)
+{
+    TextureClass *new_texture = Find_Texture(texture->Get_Name(), color);
+
+    if (new_texture) {
+        return new_texture;
+    } else {
+        return Recolor_Texture_One_Time(texture, color);
+    }
+}
+
+void Create_Color_Texture_Name(char *buffer, const char *name, int color)
+{
+    char new_name[256];
+    strcpy(new_name, name);
+    strlwr(new_name);
+    sprintf(buffer, "#%d#%s", color, new_name);
+}
+
+TextureClass *GameAssetManager::Find_Texture(const char *name, int color)
+{
+    char buffer[512];
+    Create_Color_Texture_Name(buffer, name, color);
+    TextureClass *texture = m_textureHash.Get(buffer);
+
+    if (texture) {
+        texture->Add_Ref();
+    }
+
+    return texture;
+}
+
+TextureClass *GameAssetManager::Recolor_Texture_One_Time(TextureClass *texture, int color)
+{
+    const char *name = texture->Get_Name();
+
+    if (name && name[0] == '!') {

```

Comment:
This looks odd. It makes nullptr check here, but would not early return if it did not start with exclamation mark. Then code would continue to run and eventually crash down below when `name` is dereferenced. Can Texture Name be null to begin with?
```c++
if (name == nullptr || name[0] == '!') {
```
would be more logical here.

Reply:
This is what the original code did so it must be OK otherwise it would be crashing all the time.

Reply:
I think it looks wrong and should potentially be changed. Can you add a TODO to investigate.

Reply:
By definition if the original code works and doesn't crash, it can't be "wrong".

Reply:
Compilers have improved over the years and when you write situations that can never happen it allows optimisers to delete code.